src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +174,105 @@ private sealed class RunningInstance
                 _stopping = false;
             }
 
-            private void CancellationCallback()
+            private static class StaticWatcherRunLoopManager
             {
-                lock (StopLock)
+                // A reference to the RunLoop that we can use to start or stop a Watcher
+                private static CFRunLoopRef _watcherRunLoop = IntPtr.Zero;
+
+                private static int scheduledStreamsCount = 0;
+
+                private static object lockObject = new object();
+
+                public static void ScheduleEventStream(SafeEventStreamHandle eventStream)
+                {
+                    if(_watcherRunLoop == IntPtr.Zero)
+                    {
+                        lock (lockObject)
+                        {
+                            if (_watcherRunLoop == IntPtr.Zero)
+                            {
+                                var runLoopStarted = new ManualResetEventSlim();
+                                new Thread(WatchForFileSystemEventsThreadStart) { IsBackground = true }.Start(new object[] { runLoopStarted, eventStream });
+                                runLoopStarted.Wait();
+
+                                Interlocked.Increment(ref scheduledStreamsCount);
+                                return;
+                            }
+                        }
+                    }
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                    Interlocked.Increment(ref scheduledStreamsCount);
+                }
+
+                public static void UnscheduleFromRunLoop(SafeEventStreamHandle eventStream)
+                {
+                    if (_watcherRunLoop != IntPtr.Zero)
+                    {
+                        // Always unschedule the RunLoop before cleaning up
+                        Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                        Interlocked.Decrement(ref scheduledStreamsCount);
+
+                        if (scheduledStreamsCount == 0)
```

Comment:
This isn't thread-safe.  It should be using the return value of Decrement.

Also what happens if this thread decrements to 0, goes to stop the loop, and then another thread increments and assumes it's already been started?